### PR TITLE
Ability to set vCPU weight

### PR DESCRIPTION
**Pull request by olivierlambert**

### CODE REVIEW

File: src/api/vm.coffee [hunk 1]
```diff
@@ -1007,6 +1007,25 @@ exports.setBootOrder = setBootOrder
 
 #---------------------------------------------------------------------
 
+setVcpuWeight = $coroutine ({vm, weight}) ->
+  xapi = @getXAPI vm
+
+  weight = {weight: weight}
```
Comment (julien-f) [inline src/api/vm.coffee]: Why passing an object?

File: src/api/vm.coffee [hunk 2]
```diff
@@ -1007,6 +1007,25 @@ exports.setBootOrder = setBootOrder
...
+  yield xapi.setVcpuWeight(vm._xapiRef, weight)
```
Comment (julien-f) [inline src/api/vm.coffee]: Use `vm._xapiId` instead.

File: src/xapi.js [hunk 3]
```diff
@@ -1274,6 +1274,15 @@ export default class Xapi extends XapiBase {
     )
   }
 
+  async _setVcpuWeight (vm, weight) {
+    await this._updateObjectMapProperty(vm, 'VCPUs_params', weight)
```
Comment (julien-f) [inline src/xapi.js]: `Xapi#_updateObjectMapProperty()` expects a map for the last param `{ weight }`.

File: src/api/vm.coffee [hunk 4]
```diff
@@ -1013,6 +1013,25 @@ exports.setBootOrder = setBootOrder
 
 #---------------------------------------------------------------------
 
+setVcpuWeight = $coroutine ({vm, weight}) ->
```
Comment (julien-f) [inline src/api/vm.coffee]: We should probably merge this method with `vm.set()`.

@marsaud What's your opinion?

Comment (marsaud) [inline src/api/vm.coffee]: @julien-f I agree

**Comment by olivierlambert** [comment]
I can't manage to finish efficiently the job here :( (or I need assistance to avoid losing too much time)

@marsaud can you clone this branch and make the improvements directly?

**Comment by marsaud** [comment]
I'll take care of that

**Comment by olivierlambert** [comment]
Thx!

File: src/xapi.js [hunk 5]
```diff
@@ -1657,6 +1657,21 @@ export default class Xapi extends XapiBase {
     )
   }
 
+  async _setVcpuParams (vm, params) {
```
Comment (julien-f) [inline src/xapi.js]: Not necessary.

File: src/xapi.js [hunk 6]
```diff
@@ -1657,6 +1657,21 @@ export default class Xapi extends XapiBase {
...
+    await this._updateObjectMapProperty(vm, 'VCPUs_params', params)
+  }
+
+  async setVcpuParams (vm, params) {
```
Comment (julien-f) [inline src/xapi.js]: I don't want to expose such a low level functionality. (And public methods takes ids, not objects).

File: src/xapi.js [hunk 7]
```diff
@@ -1657,6 +1657,21 @@ export default class Xapi extends XapiBase {
...
+    await this._setVcpuParams(vm, 'VCPUs_params', params)
+  }
+
+  async setVcpuWeight (vmId, weight) {
+    const vm = this.getObject(vmId)
+    const params = vm.VCPUs_params
```
Comment (julien-f) [inline src/xapi.js]: You don't need to get existing values, `Xapi#updateObjectMapProperty()` will only change specified entries.

File: src/api/vm.coffee [hunk 8]
```diff
@@ -362,6 +366,8 @@ set.params = {
 
   # Kernel arguments for PV VM.
   PV_args: { type: 'string', optional: true }
+
+  VCPUs_params: { type: 'object', optional: true}
```
Comment (julien-f) [inline src/api/vm.coffee]: `cpuWeight` and an `integer`.

**Comment by julien-f** [comment]
LGTM, clean history (squash or recommit) and I'll merge.

**Pull request merged**